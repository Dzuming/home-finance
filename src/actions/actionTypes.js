const asyncActionType = type => ({
  PENDING: `${type}_PENDING`,
  SUCCESS: `${type}_SUCCESS`,
  ERROR: `${type}_ERROR`
});

const crudActionType = type => ({
  CREATE: `CREATE_${type}`,
  READ: `READ_${type}`,
  UPDATE: `UPDATE_${type}`,
  DELETE: `DELETE_${type}`
});

const draggedElements = type => ({
  READ: `READ_${type}`,
  SELECT: `SELECT${type}`,
  RESET: `RESET_${type}`,
  REMOVE: `REMOVE_${type}`
});

export const OPEN_SIDENAV = 'OPEN_SIDENAV';
export const AUTH_USER = 'AUTH_USER';
export const AUTH_REQUEST = 'AUTH_REQUEST';
export const REQUEST_LOGOUT = 'REQUEST_LOGOUT';
export const SUCCESS_LOGOUT = 'SUCCESS_LOGOUT';
export const SET_USER = 'SET_USER';
export const SET_DATE = 'SET_DATE';
export const API_REQUEST_GET = 'API_REQUEST_GET';
export const SET_CATEGORIES = 'SET_CATEGORIES';
export const API_REQUEST_POST = 'API_REQUEST_POST';
export const API_REQUEST_DELETE = 'API_REQUEST_DELETE';
export const API_REQUEST_PUT = 'API_REQUEST_PUT';
export const API_REQUEST = asyncActionType('API_REQUEST');
export const SPENDING_CRUD = crudActionType('SPENDING');
export const PROFIT_CRUD = crudActionType('PROFIT');
export const BUDGET_CRUD = crudActionType('BUDGET');
export const REVENUE_CRUD = crudActionType('REVENUE');
export const ASSUMPTION_CRUD = crudActionType('ASSUMPTION');
export const OVERALL_ASSUMPTIONS_CRUD = crudActionType('OVERALL_ASSUMPTIONS');
export const ASSUMPTION_TYPES = draggedElements('ASSUMPTION_TYPES');
export const CATEGORIES_DRAG = draggedElements('CATEGORIES');
