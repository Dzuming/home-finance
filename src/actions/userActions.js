import * as types from './actionTypes';
import env from '../../environments/config';

function setStorageUser (user) {
  localStorage.setItem('user', JSON.stringify(user));
}

export function getUser (email, accessToken) {
  const url = `${env.api_url}/api/user/${email}`;
  return dispatch => {
    return fetch(url, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${accessToken}`,
      },
    }).then(response => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    }).then((data) => {
      dispatch(setUser(data));
    }).catch((error => error));
  };
}

export function setUser (user) {
  const newUser = {
    id: user.id,
    name: user.name,
  };
  setStorageUser(newUser);
  return {type: types.SET_USER, newUser};
}
