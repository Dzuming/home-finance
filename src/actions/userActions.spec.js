import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from './userActions';
import * as types from './actionTypes';
import fetchMock from 'fetch-mock';
import expect from 'expect';
import env from '../../environments/config';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('user actions', () => {
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('creates GET_USER when get user has been done', () => {
    const email = 'test@test.com';
    fetchMock
      .getOnce(`${env.api_url}/api/user/${email}`, {
        body: {id: '123',name: 'John Doe'},
        headers: {'content-type': 'application/json'}
      });
    const expectedActions = [
      {type: types.GET_USER, newUser: {id: '123', name: 'John Doe'}},
    ];
    const store = mockStore({user: '[]'});
    return store.dispatch(actions.getUser(email)).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
