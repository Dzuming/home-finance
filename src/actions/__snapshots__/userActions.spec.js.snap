// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user actions should create setUser action with snapshot 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "id": 1,
      "name": "Dawid",
    },
  },
  "type": "SET_USER",
}
`;
